File: lib/utils/whiteboardUtils.ts
/**
 * Whiteboard Utilities
 * Helper functions for drawing shapes, text, and canvas manipulation
 */

export type ToolType = 'pen' | 'highlighter' | 'eraser' | 'pointer' | 'text' | 'rectangle' | 'circle' | 'ellipse' | 'line' | 'arrow' | 'triangle' | 'star';

export interface Point {
  x: number;
  y: number;
}

export interface DrawAction {
  id: string;
  type: 'stroke' | 'shape' | 'text' | 'image';
  tool: ToolType;
  points?: Point[];
  color: string;
  width: number;
  opacity: number;
  fill?: boolean;
  text?: string;
  fontSize?: number;
  fontFamily?: string;
  layerId: string;
  shapeType?: 'rectangle' | 'circle' | 'ellipse' | 'line' | 'arrow' | 'triangle' | 'star';
  startPoint?: Point;
  endPoint?: Point;
  imageData?: string;
  imagePosition?: Point;
  imageSize?: { width: number; height: number };
}

export interface Layer {
  id: string;
  name: string;
  visible: boolean;
  locked: boolean;
  opacity: number;
  zIndex: number;
}

/**
 * Draw a stroke (pen, highlighter, eraser)
 */
export function drawStroke(
  ctx: CanvasRenderingContext2D,
  points: Point[],
  color: string,
  width: number,
  opacity: number,
  tool: ToolType
): void {
  if (points.length < 2) return;

  ctx.save();
  ctx.globalAlpha = opacity;
  ctx.strokeStyle = tool === 'eraser' ? '#ffffff' : color;
  ctx.lineWidth = width;
  ctx.lineCap = 'round';
  ctx.lineJoin = 'round';

  if (tool === 'highlighter') {
    ctx.globalCompositeOperation = 'multiply';
  } else if (tool === 'eraser') {
    ctx.globalCompositeOperation = 'destination-out';
  }

  ctx.beginPath();
  ctx.moveTo(points[0].x, points[0].y);

  for (let i = 1; i < points.length; i++) {
    const midPoint = {
      x: (points[i - 1].x + points[i].x) / 2,
      y: (points[i - 1].y + points[i].y) / 2,
    };
    ctx.quadraticCurveTo(points[i - 1].x, points[i - 1].y, midPoint.x, midPoint.y);
  }

  ctx.lineTo(points[points.length - 1].x, points[points.length - 1].y);
  ctx.stroke();
  ctx.restore();
}

/**
 * Draw a rectangle
 */
export function drawRectangle(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number,
  fill: boolean
): void {
  ctx.save();
  ctx.globalAlpha = opacity;

  const x = Math.min(start.x, end.x);
  const y = Math.min(start.y, end.y);
  const w = Math.abs(end.x - start.x);
  const h = Math.abs(end.y - start.y);

  if (fill) {
    ctx.fillStyle = color;
    ctx.fillRect(x, y, w, h);
  } else {
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.strokeRect(x, y, w, h);
  }

  ctx.restore();
}

/**
 * Draw a circle
 */
export function drawCircle(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number,
  fill: boolean
): void {
  ctx.save();
  ctx.globalAlpha = opacity;

  const centerX = (start.x + end.x) / 2;
  const centerY = (start.y + end.y) / 2;
  const radius = Math.sqrt(Math.pow(end.x - start.x, 2) + Math.pow(end.y - start.y, 2)) / 2;

  ctx.beginPath();
  ctx.arc(centerX, centerY, radius, 0, 2 * Math.PI);

  if (fill) {
    ctx.fillStyle = color;
    ctx.fill();
  } else {
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.stroke();
  }

  ctx.restore();
}

/**
 * Draw an ellipse
 */
export function drawEllipse(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number,
  fill: boolean
): void {
  ctx.save();
  ctx.globalAlpha = opacity;

  const centerX = (start.x + end.x) / 2;
  const centerY = (start.y + end.y) / 2;
  const radiusX = Math.abs(end.x - start.x) / 2;
  const radiusY = Math.abs(end.y - start.y) / 2;

  ctx.beginPath();
  ctx.ellipse(centerX, centerY, radiusX, radiusY, 0, 0, 2 * Math.PI);

  if (fill) {
    ctx.fillStyle = color;
    ctx.fill();
  } else {
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.stroke();
  }

  ctx.restore();
}

/**
 * Draw a line
 */
export function drawLine(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number
): void {
  ctx.save();
  ctx.globalAlpha = opacity;
  ctx.strokeStyle = color;
  ctx.lineWidth = width;
  ctx.lineCap = 'round';

  ctx.beginPath();
  ctx.moveTo(start.x, start.y);
  ctx.lineTo(end.x, end.y);
  ctx.stroke();

  ctx.restore();
}

/**
 * Draw an arrow
 */
export function drawArrow(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number
): void {
  ctx.save();
  ctx.globalAlpha = opacity;
  ctx.strokeStyle = color;
  ctx.fillStyle = color;
  ctx.lineWidth = width;
  ctx.lineCap = 'round';

  // Draw the line
  ctx.beginPath();
  ctx.moveTo(start.x, start.y);
  ctx.lineTo(end.x, end.y);
  ctx.stroke();

  // Draw the arrowhead
  const angle = Math.atan2(end.y - start.y, end.x - start.x);
  const headLength = width * 4;

  ctx.beginPath();
  ctx.moveTo(end.x, end.y);
  ctx.lineTo(
    end.x - headLength * Math.cos(angle - Math.PI / 6),
    end.y - headLength * Math.sin(angle - Math.PI / 6)
  );
  ctx.lineTo(
    end.x - headLength * Math.cos(angle + Math.PI / 6),
    end.y - headLength * Math.sin(angle + Math.PI / 6)
  );
  ctx.closePath();
  ctx.fill();

  ctx.restore();
}

/**
 * Draw a triangle
 */
export function drawTriangle(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number,
  fill: boolean
): void {
  ctx.save();
  ctx.globalAlpha = opacity;

  const x = Math.min(start.x, end.x);
  const y = Math.min(start.y, end.y);
  const w = Math.abs(end.x - start.x);
  const h = Math.abs(end.y - start.y);

  ctx.beginPath();
  ctx.moveTo(x + w / 2, y); // Top point
  ctx.lineTo(x + w, y + h); // Bottom right
  ctx.lineTo(x, y + h); // Bottom left
  ctx.closePath();

  if (fill) {
    ctx.fillStyle = color;
    ctx.fill();
  } else {
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.stroke();
  }

  ctx.restore();
}

/**
 * Draw a star
 */
export function drawStar(
  ctx: CanvasRenderingContext2D,
  start: Point,
  end: Point,
  color: string,
  width: number,
  opacity: number,
  fill: boolean
): void {
  ctx.save();
  ctx.globalAlpha = opacity;

  const centerX = (start.x + end.x) / 2;
  const centerY = (start.y + end.y) / 2;
  const outerRadius = Math.sqrt(Math.pow(end.x - start.x, 2) + Math.pow(end.y - start.y, 2)) / 2;
  const innerRadius = outerRadius * 0.4;
  const spikes = 5;

  ctx.beginPath();
  for (let i = 0; i < spikes * 2; i++) {
    const radius = i % 2 === 0 ? outerRadius : innerRadius;
    const angle = (i * Math.PI) / spikes - Math.PI / 2;
    const x = centerX + radius * Math.cos(angle);
    const y = centerY + radius * Math.sin(angle);

    if (i === 0) {
      ctx.moveTo(x, y);
    } else {
      ctx.lineTo(x, y);
    }
  }
  ctx.closePath();

  if (fill) {
    ctx.fillStyle = color;
    ctx.fill();
  } else {
    ctx.strokeStyle = color;
    ctx.lineWidth = width;
    ctx.stroke();
  }

  ctx.restore();
}

/**
 * Draw text
 */
export function drawText(
  ctx: CanvasRenderingContext2D,
  text: string,
  position: Point,
  color: string,
  fontSize: number,
  fontFamily: string,
  opacity: number
): void {
  ctx.save();
  ctx.globalAlpha = opacity;
  ctx.fillStyle = color;
  ctx.font = `${fontSize}px ${fontFamily}`;
  ctx.textBaseline = 'top';

  const lines = text.split('\n');
  lines.forEach((line, index) => {
    ctx.fillText(line, position.x, position.y + index * fontSize * 1.2);
  });

  ctx.restore();
}

/**
 * Draw an image
 */
export function drawImage(
  ctx: CanvasRenderingContext2D,
  imageElement: HTMLImageElement,
  position: Point,
  size: { width: number; height: number },
  opacity: number
): void {
  ctx.save();
  ctx.globalAlpha = opacity;
  ctx.drawImage(imageElement, position.x, position.y, size.width, size.height);
  ctx.restore();
}

/**
 * Get text dimensions
 */
export function getTextDimensions(
  ctx: CanvasRenderingContext2D,
  text: string,
  fontSize: number,
  fontFamily: string
): { width: number; height: number } {
  ctx.save();
  ctx.font = `${fontSize}px ${fontFamily}`;
  const lines = text.split('\n');
  const width = Math.max(...lines.map(line => ctx.measureText(line).width));
  const height = lines.length * fontSize * 1.2;
  ctx.restore();
  return { width, height };
}

/**
 * Clear canvas
 */
export function clearCanvas(ctx: CanvasRenderingContext2D, width: number, height: number): void {
  ctx.clearRect(0, 0, width, height);
}

/**
 * Fill canvas with background color
 */
export function fillCanvas(ctx: CanvasRenderingContext2D, width: number, height: number, color: string): void {
  ctx.save();
  ctx.fillStyle = color;
  ctx.fillRect(0, 0, width, height);
  ctx.restore();
}

/**
 * Draw grid overlay
 */
export function drawGrid(
  ctx: CanvasRenderingContext2D,
  width: number,
  height: number,
  gridSize: number,
  color: string
): void {
  ctx.save();
  ctx.strokeStyle = color;
  ctx.lineWidth = 1;
  ctx.globalAlpha = 0.2;

  // Draw vertical lines
  for (let x = 0; x <= width; x += gridSize) {
    ctx.beginPath();
    ctx.moveTo(x, 0);
    ctx.lineTo(x, height);
    ctx.stroke();
  }

  // Draw horizontal lines
  for (let y = 0; y <= height; y += gridSize) {
    ctx.beginPath();
    ctx.moveTo(0, y);
    ctx.lineTo(width, y);
    ctx.stroke();
  }

  ctx.restore();
}

/**
 * Export canvas to data URL
 */
export function exportCanvas(canvas: HTMLCanvasElement, format: 'png' | 'jpg' | 'svg' = 'png'): string {
  if (format === 'svg') {
    // For SVG, we'd need a more complex implementation
    // For now, fallback to PNG
    return canvas.toDataURL('image/png');
  }
  
  const mimeType = format === 'jpg' ? 'image/jpeg' : 'image/png';
  return canvas.toDataURL(mimeType, 0.95);
}

/**
 * Download canvas as file
 */
export function downloadCanvas(canvas: HTMLCanvasElement, filename: string, format: 'png' | 'jpg' = 'png'): void {
  const dataURL = exportCanvas(canvas, format);
  const link = document.createElement('a');
  link.download = `${filename}.${format}`;
  link.href = dataURL;
  link.click();
}

/**
 * Calculate distance between two points
 */
export function distance(p1: Point, p2: Point): number {
  return Math.sqrt(Math.pow(p2.x - p1.x, 2) + Math.pow(p2.y - p1.y, 2));
}

/**
 * Check if point is inside rectangle
 */
export function isPointInRect(point: Point, rect: { x: number; y: number; width: number; height: number }): boolean {
  return (
    point.x >= rect.x &&
    point.x <= rect.x + rect.width &&
    point.y >= rect.y &&
    point.y <= rect.y + rect.height
  );
}

/**
 * Generate unique ID
 */
export function generateId(): string {
  return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

File: almajd_app/lib/utils/whiteboard_utils.dart
/// Whiteboard drawing utilities
/// Helper functions for drawing shapes, text, and canvas manipulation

import 'dart:math' as math;
import 'dart:ui' as ui;
import 'package:flutter/material.dart';
import '../models/whiteboard_models.dart';

/// Generate unique ID for actions
String generateId() {
  final timestamp = DateTime.now().millisecondsSinceEpoch;
  final random = math.Random().nextInt(0x100000000).toRadixString(36);
  return '$timestamp-$random';
}

/// Draw a stroke (pen, highlighter, eraser) with smooth curves
void drawStroke(
  Canvas canvas,
  List<Point> points,
  Color color,
  double width,
  double opacity,
  ToolType tool,
) {
  if (points.length < 2) return;

  final paint = Paint()
    ..strokeWidth = width
    ..style = PaintingStyle.stroke
    ..strokeCap = StrokeCap.round
    ..strokeJoin = StrokeJoin.round;

  // Set blend mode and color based on tool
  if (tool == ToolType.highlighter) {
    paint.color = color.withOpacity(opacity);
    paint.blendMode = BlendMode.multiply;
  } else if (tool == ToolType.eraser) {
    // For eraser, use dstOut blend mode with white color (matches web destination-out)
    // This erases pixels instead of drawing
    paint.color = Colors.white;
    paint.blendMode = BlendMode.dstOut;
  } else {
    paint.color = color.withOpacity(opacity);
  }

  final path = Path();
  path.moveTo(points[0].x, points[0].y);

  // Use quadratic curves for smooth strokes
  for (int i = 1; i < points.length; i++) {
    final midPoint = Point(
      x: (points[i - 1].x + points[i].x) / 2,
      y: (points[i - 1].y + points[i].y) / 2,
    );
    path.quadraticBezierTo(
      points[i - 1].x,
      points[i - 1].y,
      midPoint.x,
      midPoint.y,
    );
  }

  path.lineTo(points[points.length - 1].x, points[points.length - 1].y);
  canvas.drawPath(path, paint);
}

/// Draw a rectangle
void drawRectangle(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
  bool fill,
) {
  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..style = fill ? PaintingStyle.fill : PaintingStyle.stroke
    ..strokeWidth = fill ? 0 : width;

  final rect = Rect.fromPoints(
    Offset(start.x, start.y),
    Offset(end.x, end.y),
  );

  canvas.drawRect(rect, paint);
}

/// Draw a circle
void drawCircle(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
  bool fill,
) {
  final centerX = (start.x + end.x) / 2;
  final centerY = (start.y + end.y) / 2;
  final radius = math.sqrt(
    math.pow(end.x - start.x, 2) + math.pow(end.y - start.y, 2),
  ) / 2;

  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..style = fill ? PaintingStyle.fill : PaintingStyle.stroke
    ..strokeWidth = fill ? 0 : width;

  canvas.drawCircle(Offset(centerX, centerY), radius, paint);
}

/// Draw an ellipse
void drawEllipse(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
  bool fill,
) {
  final centerX = (start.x + end.x) / 2;
  final centerY = (start.y + end.y) / 2;
  final radiusX = (end.x - start.x).abs() / 2;
  final radiusY = (end.y - start.y).abs() / 2;

  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..style = fill ? PaintingStyle.fill : PaintingStyle.stroke
    ..strokeWidth = fill ? 0 : width;

  final rect = Rect.fromCenter(
    center: Offset(centerX, centerY),
    width: radiusX * 2,
    height: radiusY * 2,
  );

  canvas.drawOval(rect, paint);
}

/// Draw a line
void drawLine(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
) {
  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..strokeWidth = width
    ..strokeCap = StrokeCap.round
    ..style = PaintingStyle.stroke;

  canvas.drawLine(
    Offset(start.x, start.y),
    Offset(end.x, end.y),
    paint,
  );
}

/// Draw an arrow with arrowhead
void drawArrow(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
) {
  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..strokeWidth = width
    ..strokeCap = StrokeCap.round
    ..style = PaintingStyle.stroke;

  // Draw the line
  canvas.drawLine(
    Offset(start.x, start.y),
    Offset(end.x, end.y),
    paint,
  );

  // Draw the arrowhead
  final angle = math.atan2(end.y - start.y, end.x - start.x);
  final headLength = width * 4;

  final arrowPath = Path();
  arrowPath.moveTo(end.x, end.y);
  arrowPath.lineTo(
    end.x - headLength * math.cos(angle - math.pi / 6),
    end.y - headLength * math.sin(angle - math.pi / 6),
  );
  arrowPath.lineTo(
    end.x - headLength * math.cos(angle + math.pi / 6),
    end.y - headLength * math.sin(angle + math.pi / 6),
  );
  arrowPath.close();

  final arrowPaint = Paint()
    ..color = color.withOpacity(opacity)
    ..style = PaintingStyle.fill;

  canvas.drawPath(arrowPath, arrowPaint);
}

/// Draw a triangle
void drawTriangle(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
  bool fill,
) {
  final x = math.min(start.x, end.x);
  final y = math.min(start.y, end.y);
  final w = (end.x - start.x).abs();
  final h = (end.y - start.y).abs();

  final path = Path();
  path.moveTo(x + w / 2, y); // Top point
  path.lineTo(x + w, y + h); // Bottom right
  path.lineTo(x, y + h); // Bottom left
  path.close();

  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..style = fill ? PaintingStyle.fill : PaintingStyle.stroke
    ..strokeWidth = fill ? 0 : width;

  canvas.drawPath(path, paint);
}

/// Draw a star
void drawStar(
  Canvas canvas,
  Point start,
  Point end,
  Color color,
  double width,
  double opacity,
  bool fill,
) {
  final centerX = (start.x + end.x) / 2;
  final centerY = (start.y + end.y) / 2;
  final outerRadius = math.sqrt(
    math.pow(end.x - start.x, 2) + math.pow(end.y - start.y, 2),
  ) / 2;
  final innerRadius = outerRadius * 0.4;
  const spikes = 5;

  final path = Path();
  for (int i = 0; i < spikes * 2; i++) {
    final radius = i % 2 == 0 ? outerRadius : innerRadius;
    final angle = (i * math.pi) / spikes - math.pi / 2;
    final x = centerX + radius * math.cos(angle);
    final y = centerY + radius * math.sin(angle);

    if (i == 0) {
      path.moveTo(x, y);
    } else {
      path.lineTo(x, y);
    }
  }
  path.close();

  final paint = Paint()
    ..color = color.withOpacity(opacity)
    ..style = fill ? PaintingStyle.fill : PaintingStyle.stroke
    ..strokeWidth = fill ? 0 : width;

  canvas.drawPath(path, paint);
}

/// Draw text on canvas
void drawText(
  Canvas canvas,
  String text,
  Point position,
  Color color,
  double fontSize,
  String fontFamily,
  double opacity,
) {
  final textStyle = TextStyle(
    color: color.withOpacity(opacity),
    fontSize: fontSize,
    fontFamily: fontFamily,
  );

  final textSpan = TextSpan(text: text, style: textStyle);
  final textPainter = TextPainter(
    text: textSpan,
    textDirection: TextDirection.ltr,
    textAlign: TextAlign.left,
  );
  textPainter.layout();

  final lines = text.split('\n');
  for (int i = 0; i < lines.length; i++) {
    final lineSpan = TextSpan(text: lines[i], style: textStyle);
    final linePainter = TextPainter(
      text: lineSpan,
      textDirection: TextDirection.ltr,
    );
    linePainter.layout();
    linePainter.paint(
      canvas,
      Offset(position.x, position.y + i * fontSize * 1.2),
    );
  }
}

/// Draw an image on canvas
Future<void> drawImage(
  Canvas canvas,
  ui.Image image,
  Point position,
  ImageSize size,
  double opacity,
) async {
  final paint = Paint()..color = Colors.white.withOpacity(opacity);

  canvas.drawImageRect(
    image,
    Rect.fromLTWH(0, 0, image.width.toDouble(), image.height.toDouble()),
    Rect.fromLTWH(position.x, position.y, size.width, size.height),
    paint,
  );
}

/// Draw grid overlay
void drawGrid(
  Canvas canvas,
  double width,
  double height,
  double gridSize,
  Color color,
) {
  final paint = Paint()
    ..color = color.withOpacity(0.2)
    ..strokeWidth = 1;

  // Draw vertical lines
  for (double x = 0; x <= width; x += gridSize) {
    canvas.drawLine(
      Offset(x, 0),
      Offset(x, height),
      paint,
    );
  }

  // Draw horizontal lines
  for (double y = 0; y <= height; y += gridSize) {
    canvas.drawLine(
      Offset(0, y),
      Offset(width, y),
      paint,
    );
  }
}

/// Calculate distance between two points
double distance(Point p1, Point p2) {
  return math.sqrt(
    math.pow(p2.x - p1.x, 2) + math.pow(p2.y - p1.y, 2),
  );
}

/// Check if point is inside rectangle
bool isPointInRect(
  Point point,
  Rect rect,
) {
  return point.x >= rect.left &&
      point.x <= rect.right &&
      point.y >= rect.top &&
      point.y <= rect.bottom;
}

/// Convert hex color string to Color
Color hexToColor(String hexString) {
  final buffer = StringBuffer();
  if (hexString.length == 6 || hexString.length == 7) {
    buffer.write('ff'); // Add alpha if missing
  }
  buffer.write(hexString.replaceFirst('#', ''));
  return Color(int.parse(buffer.toString(), radix: 16));
}

/// Convert Color to hex string
String colorToHex(Color color) {
  return '#${color.value.toRadixString(16).substring(2).toUpperCase()}';
}

/// Parse color from dynamic value (handles both string and int)
Color parseColor(dynamic colorData) {
  if (colorData is String) {
    if (colorData.startsWith('#')) {
      return hexToColor(colorData);
    }
    // Handle color names
    switch (colorData.toLowerCase()) {
      case 'black':
        return Colors.black;
      case 'red':
        return Colors.red;
      case 'green':
        return Colors.green;
      case 'blue':
        return Colors.blue;
      case 'yellow':
        return Colors.yellow;
      case 'orange':
        return Colors.orange;
      case 'purple':
        return Colors.purple;
      case 'pink':
        return Colors.pink;
      case 'cyan':
        return Colors.cyan;
      case 'brown':
        return Colors.brown;
      default:
        return Colors.black;
    }
  } else if (colorData is int) {
    return Color(colorData);
  }
  return Colors.black;
}

